cpp/src/arrow/json/from_string.h

```diff
@@ -0,0 +1,121 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Implement a simple JSON representation format for arrays
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include "arrow/status.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+class DataType;
+
+namespace json {
+
+/// \defgroup array-from-json-string FromJSONString Helpers
+///
+/// These helpers are intended to be used in examples, tests, or for quick
+/// prototyping and are not intended to be used where performance matters.
+///
+/// See the <a href="../arrays.html#fromjsonstring-helpers">User Guide</a> for
+/// more information.
+///
+/// @{
+
+/// \brief Create an Array from a JSON string
+///
+/// \code {.cpp}
+/// std::shared_ptr<Array> array = ArrayFromJSONString(
+///   int64(), "[2, 3, null, 7, 11]"
+/// ).ValueOrDie();
+/// \endcode
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> ArrayFromJSONString(const std::shared_ptr<DataType>&,
+                                                   const std::string& json);
+
+/// \copydoc ArrayFromJSONString(const std::shared_ptr<DataType>&, const std::string&)
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> ArrayFromJSONString(const std::shared_ptr<DataType>&,
+                                                   std::string_view json);
+
+/// \copydoc ArrayFromJSONString(const std::shared_ptr<DataType>&, const std::string&)
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> ArrayFromJSONString(const std::shared_ptr<DataType>&,
+                                                   const char* json);
+
+/// \brief Create a ChunkedArray from a JSON string
+///
+/// \code {.cpp}
+/// std::shared_ptr<ChunkedArray> chunked_array;
+/// ChunkedArrayFromJSONString(
+///   int64(), {R"([5, 10])", R"([null])", R"([16])"}, &chunked_array
+/// );
+/// \endcode
+ARROW_EXPORT
+Status ChunkedArrayFromJSONString(const std::shared_ptr<DataType>& type,
```

Comment:
Why not return a `Result<std::shared_ptr<ChunkedArray>>` instead of taking an out-pointer parameter?

Reply:
This was easy to do, I put up [a draft PR](https://github.com/apache/arrow/pull/46447) and converted ChunkedArrayFromJSONString in https://github.com/apache/arrow/pull/46447/commits/ede205e81ee19199bef512656f90cc11417ca39b.

I can do the rest if you think that makes sense (I do).

Reply:
Yes, I think we should convert all of them.

Reply:
Converted by PR https://github.com/apache/arrow/pull/46696.